public/pages/AnomalyCharts/utils/anomalyChartUtils.ts

```diff
@@ -138,3 +150,281 @@ export const disabledHistoryAnnotations = (
+    return ANOMALY_HEATMAP_COLORSCALE[0][1];
+  }
+
+  for (let i = 0; i < ANOMALY_HEATMAP_COLORSCALE.length - 1; i++) {
+    if (
+      value >= ANOMALY_HEATMAP_COLORSCALE[i][0] &&
+      value < ANOMALY_HEATMAP_COLORSCALE[i + 1][0]
+    ) {
+      return ANOMALY_HEATMAP_COLORSCALE[i + 1][1];
+    }
+  }
+};
+
+const NUM_CELLS = 20;
+
+export const HEATMAP_X_AXIS_DATE_FORMAT = 'MM-DD HH:mm YYYY';
+
+const buildBlankStringWithLength = (length: number) => {
+  let result = '';
+  for (let i = 0; i < length; i++) {
+    result += ' ';
+  }
+  return result;
+};
+
+export const getAnomaliesHeatmapData = (
+  anomalies: any[],
+  dateRange: DateRange,
+  sortType: AnomalyHeatmapSortType = AnomalyHeatmapSortType.SEVERITY,
+  displayTopNum: number
+): PlotData[] => {
+  const entityAnomaliesMap = getEntityAnomaliesMap(anomalies);
+  if (isEmpty(entityAnomaliesMap)) {
+    // put placeholder data so that heatmap won't look empty
+    for (let i = 0; i < displayTopNum; i++) {
+      const blankStrValue = buildBlankStringWithLength(i);
+      entityAnomaliesMap.set(blankStrValue, []);
+    }
+  }
+
+  const entityValues = [] as string[];
+  const maxAnomalyGrades = [] as any[];
+  const numAnomalyGrades = [] as any[];
+
+  const timeWindows = calculateTimeWindowsWithMaxDataPoints(
+    NUM_CELLS,
+    dateRange
+  );
+
+  entityAnomaliesMap.forEach((entityAnomalies, entity) => {
+    const maxAnomalyGradesForEntity = [] as number[];
+    const numAnomalyGradesForEntity = [] as number[];
+    entityValues.push(entity);
+    timeWindows.forEach((timeWindow) => {
+      const anomaliesInWindow = entityAnomalies.filter(
+        (anomaly) =>
+          get(anomaly, 'plotTime', 0) <= timeWindow.endDate &&
+          get(anomaly, 'plotTime', 0) >= timeWindow.startDate
+      );
+      const entityAnomalyGrades = anomaliesInWindow.map((anomaly) =>
+        get(anomaly, 'anomalyGrade', 0)
+      );
+      if (!isEmpty(entityAnomalyGrades)) {
+        maxAnomalyGradesForEntity.push(Math.max(...entityAnomalyGrades));
+        numAnomalyGradesForEntity.push(
+          entityAnomalyGrades.filter((anomalyGrade) => anomalyGrade > 0).length
+        );
+      } else {
+        maxAnomalyGradesForEntity.push(0);
+        numAnomalyGradesForEntity.push(0);
+      }
+    });
+
+    maxAnomalyGrades.push(maxAnomalyGradesForEntity);
+    numAnomalyGrades.push(numAnomalyGradesForEntity);
+  });
+
+  const plotTimes = timeWindows.map((timeWindow) => timeWindow.endDate);
+  const plotData =
+    //@ts-ignore
+    {
+      x: plotTimes.map((timestamp) =>
+        moment(timestamp).format(HEATMAP_X_AXIS_DATE_FORMAT)
+      ),
+      y: entityValues,
+      z: maxAnomalyGrades,
+      colorscale: ANOMALY_HEATMAP_COLORSCALE,
+      //@ts-ignore
+      zmin: 0,
+      zmax: 1,
+      type: 'heatmap',
+      showscale: false,
+      xgap: 2,
+      ygap: 2,
+      opacity: 1,
+      text: numAnomalyGrades,
+      hovertemplate:
+        '<b>Time</b>: %{x}<br>' +
```

Comment:
minor / future improvement: could the hover time show the date range for the selected cell instead? That could help the user understand that each cell is still an aggregation over a given time frame

Reply:
It is hard to implement that. I didn;t find solution. Original UX has such design to show time range, but due to limitation of Plotly library, we can only have 4 fields(x, y, z, text) of `PlotData` of each cell shown in [hovertemplate](https://plotly.com/javascript/reference/heatmap/#heatmap-hovertemplate). And `x` must be timestamp of string, there is no good way to customize/format any value to date string for `x`.

Reply:
Yeah, i see, not a problem, just wanted to confirm there was no easy way.